discovery-infra/test_infra/helper_classes/hypershift.py

```diff
@@ -0,0 +1,93 @@
+import subprocess
+from base64 import b64decode
+
+import waiting
+from kubernetes.client import ApiClient, CoreV1Api, CustomObjectsApi, V1NodeList
+from test_infra import utils
+from test_infra.helper_classes.kube_helpers import Secret, create_kube_api_client
+
+logger = logging.getLogger(__name__)
+
+HYPERSHIFT_DIR = "capi/hypershift/"
+DEFAULT_WAIT_FOR_NODES_TIMEOUT = 5 * 60
+
+
+class HyperShift:
+    """
+    Hypershift resource that allow interaction with hypershift CLI, nodepool and hypershift cluster kubeapi-server.
+    """
+
+    def __init__(self, name: str):
+        self.name = name
+        self.kubeconfig_path = ""
+        self.hypershift_cluster_client = None
+
+    def create(self, pull_secret_file: str, ssh_key: str = ""):
+        # TODO: change this to create namespace, pull-secret, hostedCluster and nodePool
+        logger.info(f"Creating HyperShift cluster {self.name}")
+        cmd = f"./bin/hypershift create cluster agent --pull-secret {pull_secret_file} --name {self.name}"
+        if ssh_key:
+            cmd += f" --ssh-key {ssh_key}"
+        subprocess.check_call(cmd, shell=True, cwd=HYPERSHIFT_DIR)
+
+    def delete(self):
+        logger.info(f"Deleting HyperShift cluster {self.name}")
+        subprocess.check_call(
+            f"./bin/hypershift destroy cluster agent --name {self.name}",
+            shell=True,
+            cwd=HYPERSHIFT_DIR,
+        )
+
+    def download_kubeconfig(self, kube_api_client: ApiClient) -> str:
+        logger.info(f"Downloading kubeconfig for HyperShift cluster {self.name}")
+        kubeconfig_data = (
+            Secret(
+                kube_api_client=kube_api_client,
+                namespace=f"clusters-{self.name}",
+                name="admin-kubeconfig",
+            )
+            .get()
+            .data["kubeconfig"]
+        )
+        hypershift_kubeconfig_path = utils.get_kubeconfig_path(self.name) + "-hypershift"
+
+        with open(hypershift_kubeconfig_path, "wt") as kubeconfig_file:
+            kubeconfig_file.write(b64decode(kubeconfig_data).decode())
+            kubeconfig_file.flush()
+        self.kubeconfig_path = hypershift_kubeconfig_path
+        return self.kubeconfig_path
+
+    def set_nodepool_node_count(self, kube_api_client: ApiClient, node_count: int) -> None:
+        logger.info(f"Setting HyperShift cluster {self.name} node count to: {node_count}")
+        crd_api = CustomObjectsApi(kube_api_client)
+        node_count = node_count
+        body = {"spec": {"nodeCount": node_count}}
+        crd_api.patch_namespaced_custom_object(
+            group="hypershift.openshift.io",
+            version="v1alpha1",
+            plural="nodepools",
+            name=self.name,
+            namespace="clusters",
+            body=body,
+        )
+
+    def get_nodes(self) -> V1NodeList:
+        if self.hypershift_cluster_client is None:
+            hypershift_cluter_kubeapi_client = create_kube_api_client(self.kubeconfig_path)
```

Comment:
1. See the line above (76) it only get created if `self.hypershift_cluster_client is None`.
2. We can't create it in init because the kubeconfig for it get created in a  later stage.